### PR TITLE
Fix read name hashing.

### PR DESCRIPTION
**Pull request by kvg**
* Fixed read name hashing by keeping an incrementing index as reads are ingested and dispatched to threads.
* Fixed a bug resulting from renaming a function.

### CODE REVIEW

File: src/longbow/segment/command.py [hunk 1]
```diff
@@ -526,11 +537,14 @@ def _write_segmented_read(
...
     :param bam_out: An open pysam.AlignmentFile ready to write out data.
     :param barcode_conf_file: An open file ready to write out the barcodes and confidence scores.
+    :param ignore_cbc_and_umi: Boolean indicating whether we should ignore CBC/UMI tags.
+    :param read_index: The index of the read.
     :return: the number of segments written.
     """
 
     segment_bounds_tuples = segment_read_with_simple_splitting(read, delimiters, segments)
 
+    split_read_index = read_index * 100
```
Comment (jonn-smith) [inline src/longbow/segment/command.py:547]: I think we should make this dependent on the number of segments in the model.  If we somehow had a lot of model segments, 100 might not be big enough to disambiguate them all (it's good for the 15 element ones).

File: src/longbow/utils/bam_utils.py [hunk 2]
```diff
@@ -418,6 +418,11 @@ def get_model_name_from_bam_header(header):
     return get_model_name_from_bam_header(header)['name']
 
 
+def get_model_from_bam_header(header):
+    model_jsons = get_models_from_bam_header(header)
+    return model_jsons[0]
```
Comment (jonn-smith) [inline src/longbow/utils/bam_utils.py:423]: Can you add a warning or error if there is more than one model in the header?

**Review by jonn-smith — CHANGES_REQUESTED** [review]
A couple of quick things and one question:

As is, won't this produce the same read names for (at least) the very first read in multiple files?  I think there needs to be some additional disambiguation based on an additional marker (read content, length, aligned position, something else).

File: src/longbow/segment/command.py (hunk 1 above)
Comment (kvg) [reply src/longbow/segment/command.py:547]: Good idea - will change.

File: src/longbow/utils/bam_utils.py (hunk 2 above)
Comment (kvg) [reply src/longbow/utils/bam_utils.py:423]: Will do.

**Comment by kvg** [comment]
Requested changes now made.

**Comment by kvg** [comment]
> A couple of quick things and one question:
> 
> As is, won't this produce the same read names for (at least) the very first read in multiple files? I think there needs to be some additional disambiguation based on an additional marker (read content, length, aligned position, something else).

Oh wait, you're right. I forgot to handle this case.  Let me think about this some more.

**Comment by kvg** [comment]
Okay, I think I've come up with a scheme that works. This works off the ZMW number. I pad the ZMW number, pad the split read index, and concatenate the two. I've written a test to exhaustively demonstrate that in up to 8,000,000 ZMW numbers and up to 20 array elements, the new read names never collide.

File: src/longbow/segment/command.py [hunk 3]
```diff
@@ -526,11 +528,14 @@ def _write_segmented_read(
     :param delimiters: A list of tuples containing the names of delimiter sequences to use to split the given read.
     :param bam_out: An open pysam.AlignmentFile ready to write out data.
     :param barcode_conf_file: An open file ready to write out the barcodes and confidence scores.
+    :param ignore_cbc_and_umi: Boolean indicating whether we should ignore CBC/UMI tags.
+    :param split_read_index: The index of the read.
```
Comment (jonn-smith) [inline src/longbow/segment/command.py:532]:
```suggestion
```

File: src/longbow/segment/command.py [hunk 4]
```diff
@@ -216,8 +216,10 @@ def _sub_process_work_fn(in_queue, out_queue):
             return
 
         # Unpack our data here:
+        read = raw_data
```
Comment (jonn-smith) [inline src/longbow/segment/command.py:219]: This doesn't seem to do anything - probably faster to just use `raw_data` in the `fromstring` construction below.

**Review by jonn-smith — APPROVED** [review]
Looks good!  Two quick suggestions then feel free to merge.

File: src/longbow/segment/command.py (hunk 4 above)
Comment (kvg) [reply src/longbow/segment/command.py:219]: Good point, will change.

**Pull request merged**